### PR TITLE
feat(context-headers): Allow passing of context headers to signing request [EXT-2835]

### PR DESCRIPTION
**Pull request by kdamball**
Allow request signing to provide additional headers for user context

### CODE REVIEW

File: src/requests/sign-request.ts [hunk 1]
```diff
@@ -102,7 +104,8 @@ const getSortedAndSignedHeaders = (headers: Record<string, string>, timestamp: n
 export const signRequest = (
   rawSecret: Secret,
   rawCanonicalRequest: CanonicalRequest,
-  rawTimestamp: Timestamp = Date.now()
+  rawTimestamp: Timestamp = Date.now(),
+  contextHeaders: Record<string, string> = {}
```
Comment (Jwhiles) [inline src/requests/sign-request.ts:108]: maybe put a stricter type on this

File: src/requests/sign-request.ts [hunk 2]
```diff
@@ -113,11 +117,22 @@ export const signRequest = (
   const headers = canonicalRequest.headers ? normalizeHeaders(canonicalRequest.headers) : {}
   const body = canonicalRequest.body ?? ''
 
-  const { sortedHeaders, signedHeaders } = getSortedAndSignedHeaders(headers, timestamp)
+  const { sortedHeaders, signedHeaders } = getSortedAndSignedHeaders(
+    { ...headers, ...((contextHeaders as unknown) as Record<string, string>) },
```
Comment (shikaan) [inline src/requests/sign-request.ts:121]: We would need to normalize context headers as we do for the rest to make sure signing works consistently. Maybe we can do as we do above so we don't need to force the type coercion here (if I am understanding the reason for `as` correctly)

File: src/requests/sign-request.ts [hunk 3]
```diff
@@ -113,11 +117,22 @@ export const signRequest = (
...
+    timestamp
+  )
+  const subject = contextHeaders[ContentfulUserIdHeader]
+    ? { [ContentfulUserIdHeader]: contextHeaders[ContentfulUserIdHeader] }
+    : contextHeaders[ContentfulAppIdHeader]
+    ? { [ContentfulAppIdHeader]: contextHeaders[ContentfulAppIdHeader] }
+    : {}
```
Comment (shikaan) [inline src/requests/sign-request.ts:128]: nit: a gool ol' if-else wrapped in a function would be much easier to read for me

File: src/requests/verify-request.spec.ts [hunk 4]
```diff
@@ -54,6 +67,36 @@ describe('isVerifiedRequest', () => {
...
+  it('verifies a verified request with user-id', () => {
+    const now = Date.now()
+    const incomingRequest = makeIncomingRequest(
+      {
+        headers: {
+          Authorization: 'Bearer TOKEN',
+        },
+      },
+      now,
+      { [ContentfulUserIdHeader]: 'my-user' }
```
Comment (shikaan) [inline src/requests/verify-request.spec.ts:79]: Segueing comments above, I would also test the case where a non-normalized header comes in

File: src/requests/sign-request.ts [hunk 5]
```diff
@@ -102,7 +105,8 @@ const getSortedAndSignedHeaders = (headers: Record<string, string>, timestamp: n
 export const signRequest = (
```
Comment (shikaan) [inline src/requests/sign-request.ts:105]: We need to update the jsdoc too here

File: src/requests/typings.ts [hunk 6]
```diff
@@ -74,4 +88,11 @@ export type NormalizedCanonicalRequest = {
   body: CanonicalRequest['body']
 }
 
-export type SignedRequestHeaders = { [key in ContentfulHeader]: string }
+export type Subject = XOR<ContentfulHeaderWithApp, ContentfulHeaderWithUser>
+
+export type ContextHeaders = {
```
Comment (Jwhiles) [inline src/requests/typings.ts:93]: I think this type is incorrect actually. As it doesn't allow the possibility of not having a subject - which is a case we want to allow.

Currently if you try

```
      signRequest(VALID_SECRET, VALID_REQUEST, VALID_TIMESTAMP, {
        'x-contentful-space-id': 'whatever',
        'x-contentful-environment-id': 'whatever',
      })
```

we get a typescript error

File: src/requests/typings.ts [hunk 7]
```diff
@@ -2,11 +2,25 @@
...
+export const ContentfulUserIdHeader = 'x-contentful-user-id'
+export const ContentfulAppIdHeader = 'x-contentful-app-id'
+
+type ContentfulHeaderWithApp = {
+  [ContentfulAppIdHeader]: string
+}
+
+type ContentfulHeaderWithUser = {
+  [ContentfulUserIdHeader]: string
 }
```
Comment (shikaan) [inline src/requests/typings.ts:24]: Why can't we unify this under `x-contentful-subject-id`?

File: src/requests/sign-request.ts (hunk 2 above)
Comment (kdamball) [reply src/requests/sign-request.ts:121]: Key thing to keep in mind here is to guarantee contextHeaders values found in `headers` are overwritten. Basically merging two objects and we trust the second one more than the other

File: src/requests/utils.ts [hunk 8]
```diff
@@ -26,6 +33,25 @@ export const pickHeaders = (headers?: Record<string, string>, keys?: string[]) =
   return filter(headers, ([key]) => keys.includes(key))
 }
 
+export const normalizeContextHeaders = (
+  rawContextHeaders: ContextHeaders
+): Partial<SignedRequestHeaders> => {
+  const contextHeadersMap: Record<string, string> = {
```
Comment (shikaan) [inline src/requests/utils.ts:39]: nit: this might be defined outside to not recreate it every time

**Review by shikaan — APPROVED** [review]

**Pull request merged**